Adding a Kanister function to create CSI VolumeSnapshot (#1163)
kanisterio/kanister: docs/functions.rst

                  name: http
                  protocol: TCP

> 

Reviewer: Please add this line back. It's required for proper formatting of the docs

              resource: namespaces
              name: "{{ .Namespace.Name }}"

> CreateCSISnapshot

Reviewer: ```suggestion

CreateCSISnapshot
```
Two new lines are required between different functions for formatting

-----------------

> This function is used to Create CSI VolumeSnapshot for a PersistentVolumeClaim.
> It halts the action until ``ReadyToUse`` status of VolumeSnapshot is ``true``.

Reviewer: ```suggestion
This function is used to create CSI VolumeSnapshot for a PersistentVolumeClaim.
By default, it waits for the ``ReadyToUse`` status of the VolumeSnapshot to be ``true``.
```

Arguments:

.. csv-table::
   :header: "Argument", "Required", "Type", "Description"
   :align: left
   :widths: 5,5,5,15

>    `name`, Yes, `string`, name of the VolumeSnapshot

Reviewer: As suggested below, let's make this an optional arg

   `namespace`, Yes, `string`, namespace of the PersistentVolumeClaim and resultant VolumeSnapshot
   `snapshotClass`, Yes, `string`, name of the VolumeSnapshotClass
   `labels`, No, `map[string]string`, labels for the VolumeSnapshot

Outputs:

.. csv-table::
   :header: "Output", "Type", "Description"
   :align: left
   :widths: 5,5,15

   `name`,`string`, name of the CSI VolumeSnapshot
   `pvc`,`string`, name of the captured PVC
   `namespace`, string, namespace of the captured PVC and VolumeSnapshot
   `restoreSize`, string, required memory size to restore PVC
   `snapshotContent`, string, name of the VolumeSnapshotContent

Example:

.. code-block:: yaml
  :linenos:

  actions:
  backup:
    outputArtifacts:
      snapshotInfo:
        keyValue:
          name: "{{ .Phases.createCSISnapshot.Output.name }}"
          pvc: "{{ .Phases.createCSISnapshot.Output.pvc }}"
          namespace: "{{ .Phases.createCSISnapshot.Output.namespace }}"
          restoreSize: "{{ .Phases.createCSISnapshot.Output.restoreSize }}"
          snapshotContent: "{{ .Phases.createCSISnapshot.Output.snapshotContent }}"
    phases:
    - func: CreateCSISnapshot
      name: createCSISnapshot
      args:

>         name: "{{ .PVC.Name }}-volumesnapshot"

Reviewer: Elaborating the point below: if a user wants to keep executing the blueprint once every hour for the same PVC, this will try creating a volume snapshot with the same name which probably will fail. Having the function generate the name will help in such scenarios.
In case, someone wants custom names, they can always modify their blueprint to specify one

Author: Okay. I'll make this update.....

              name: "{{ .Namespace.Name }}"

CreateCSISnapshot
-----------------

This function is used to create CSI VolumeSnapshot for a PersistentVolumeClaim.

> By default, it waits for the ``ReadyToUse`` status of the VolumeSnapshot to be ``true``.

Reviewer 2: ```suggestion
By default, it waits for the VolumeSnapshot to be  ready to use after creating it.
```
or something in the same lines.

Arguments:

.. csv-table::
   :header: "Argument", "Required", "Type", "Description"
   :align: left
   :widths: 5,5,5,15

   `name`, No, `string`, name of the VolumeSnapshot. Default value is ``<pvc>-snapshot-<random-alphanumeric-suffix>``
   `pvc`, Yes, `string`, name of the PersistentVolumeClaim to be captured
   `namespace`, Yes, `string`, namespace of the PersistentVolumeClaim and resultant VolumeSnapshot
   `snapshotClass`, Yes, `string`, name of the VolumeSnapshotClass
   `labels`, No, `map[string]string`, labels for the VolumeSnapshot

Outputs:

.. csv-table::
   :header: "Output", "Type", "Description"
   :align: left
   :widths: 5,5,15

   `name`,`string`, name of the CSI VolumeSnapshot
   `pvc`,`string`, name of the captured PVC
   `namespace`, string, namespace of the captured PVC and VolumeSnapshot
   `restoreSize`, string, required memory size to restore PVC
   `snapshotContent`, string, name of the VolumeSnapshotContent

Example:

.. code-block:: yaml
  :linenos:

  actions:

>   backup:
>     outputArtifacts:
>       snapshotInfo:
>         keyValue:
>           name: "{{ .Phases.createCSISnapshot.Output.name }}"
>           pvc: "{{ .Phases.createCSISnapshot.Output.pvc }}"
>           namespace: "{{ .Phases.createCSISnapshot.Output.namespace }}"
>           restoreSize: "{{ .Phases.createCSISnapshot.Output.restoreSize }}"
>           snapshotContent: "{{ .Phases.createCSISnapshot.Output.snapshotContent }}"
>     phases:
>     - func: CreateCSISnapshot
>       name: createCSISnapshot
>       args:
>         pvc: "{{ .PVC.Name }}"
>         namespace: "{{ .PVC.Namespace }}"
>         snapshotClass: do-block-storage

Reviewer 3: ```suggestion
    backup:
      outputArtifacts:
        snapshotInfo:
          keyValue:
            name: "{{ .Phases.createCSISnapshot.Output.name }}"
            pvc: "{{ .Phases.createCSISnapshot.Output.pvc }}"
            namespace: "{{ .Phases.createCSISnapshot.Output.namespace }}"
            restoreSize: "{{ .Phases.createCSISnapshot.Output.restoreSize }}"
            snapshotContent: "{{ .Phases.createCSISnapshot.Output.snapshotContent }}"
      phases:
      - func: CreateCSISnapshot
        name: createCSISnapshot
        args:
          pvc: "{{ .PVC.Name }}"
          namespace: "{{ .PVC.Namespace }}"
          snapshotClass: do-block-storage
```
fix indentation

   :align: left
   :widths: 5,5,5,15

>    `name`, No, `string`, name of the VolumeSnapshot. Default value is ``<pvc>-snapshot-<30-char-random-alphanumeric-suffix>``

Reviewer 2: Like mentioned at other place 30 is a bit too much. If this is how k8s does recommends is then its ok.

Author: K8s has a limit of 253 chars
